ttnn/cpp/ttnn/operations/eltwise/binary/binary_compiler_interface.hpp

```diff
@@ -0,0 +1,64 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <exception>
+#include <functional>
+
+#include "third_party/json/json.hpp"
+#include "tt_metal/common/logger.hpp"
+#include "ttnn/compiler_interface/compiler_interface.hpp"
+#include "ttnn/graph/graph_processor.hpp"
+#include "ttnn/graph/graph_trace_utils.hpp"
+
+namespace ttnn::compiler_interface {
+
+template <auto BinaryFunction>
+QueryResponse binary_op_constraints(
+    Device* device, const TensorSpec& input_a, const TensorSpec& input_b, const TensorSpec& output) {
+    // get_op_trace is a lambda that prepares input and output tensors, capturing graph trace of the op without
+    // inputs allocation.
+    auto get_op_trace = [](Device* device,
+                           const TensorSpec& input_a,
+                           const TensorSpec& input_b,
+                           const TensorSpec& output) {
+        nlohmann::json op_trace;
+
+        // outer graph capture is used to avoid capturing and dispatching of dummy input tensor(s) creation
+        {
+            ttnn::graph::GraphProcessor::begin_graph_capture(ttnn::graph::GraphProcessor::RunMode::NO_DISPATCH);
+            const auto input_tensor_a = create_device_tensor(input_a, device);
+            const auto input_tensor_b = create_device_tensor(input_b, device);
+
+            // output tensor is created in the inner graph capture to capture its allocation
+            {
```

Comment:
Why the extra scope?

Reply:
readability - in my mind makes a clear cut on what's captured in the inner graph capture. I can remove it if you care?

Reply:
This went away after refactoring